File: 400/454 - 4Sum II.java
import java.util.HashMap;

public class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        int sum, ret = 0;
        
        for (int a : A) {
            for (int b : B) {
                sum = a + b;
                if (hm.containsKey(sum))
                    hm.replace(sum, hm.get(sum) + 1);
                else
                    hm.put(sum, 1);
            }
        }
        
        for (int c: C) {
            for (int d : D) {
                sum = -1 * (c + d);
                
                if (hm.containsKey(sum))
                    ret += hm.get(sum);
            }
        }
        
        return ret;
    }
}

File: csharp/454.4-sum-ii.cs
/*
 * @lc app=leetcode id=454 lang=csharp
 *
 * [454] 4Sum II
 */

using System.Collections.Generic;

public class Solution {
    public int FourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int n = A.Length, ret = 0;
        IDictionary<int, int> count = new Dictionary<int, int>();
        for (int i = 0; i < n; ++i) {
            int curr = A[i];
            for (int j = 0; j < n; ++j) {
                int sum = curr + B[j];
                count[sum] = 1 + (count.ContainsKey(sum) ? count[sum] : 0);
            }
        }
        for (int i = 0; i < n; ++i) {
            int curr = C[i];
            for (int j = 0; j < n; ++j) {
                int sum = curr + D[j];
                if (count.ContainsKey(-sum)) {
                    ret += count[-sum];
                }
            }
        }
        return ret;
    }
}

